### PR TITLE
sql: disable comparison of negative int to oid

### LINKED ISSUE #60533: v20.2.2 negative int to oid cast semantics makes the jdbc driver unhappy
A customer upgrading their v20.1 to v20.2 ran into hanging queries when connecting to their cluster using `Hibernate` and `IntelliJ`.

These tools use a JDBC query to perform a check on some pg trigger before using the queries, that we believe is leading to the rest of the queries on this cluster hanging.

The query appears to succeed on 20.2 but return zero results, something that actually returns one row in postgres (verified with pg 8.3.23). This seems to be where we differ from postgres because we don't support these triggers and some other pg catalog tables.

I believe that we are able to move past them in v20.1 and prior because of the unsupported oid to integer conversion that we fixed in 20.2 as a result of this pgjdbc test https://github.com/pgjdbc/pgjdbc/blob/01dbbac5b6c43e1079a463119e13c1bee74cbfdb/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSetMetaData.java#L187.

The details of the drivers are below:
```
IntelliJ 2020.1 & 2020.3.1 Ultimate Edition
Runtime Version: 11.0.6
JDBC Driver Version: PostgreSQL - ver. 42.2.5 (latest)
Hibernate: 5.3.10.Final
```

The queries that are issued are:

`SELECT c.oid, a.attnum, a.attname, c.relname, n.nspname, a.attnotnull OR ((t.typtype = 'd') AND t.typnotnull), pg_get_expr(d.adbin, d.adrelid) LIKE '%nextval(%' FROM pg_catalog.pg_class AS c JOIN pg_catalog.pg_namespace AS n ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute AS a ON (c.oid = a.attrelid) JOIN pg_catalog.pg_type AS t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef AS d ON ((d.adrelid = a.attrelid) AND (d.adnum = a.attnum)) JOIN (SELECT 2699 AS oid, 1 AS attnum UNION ALL SELECT 2699, 2 UNION ALL SELECT 2699, 3 UNION ALL SELECT 2699, 4 UNION ALL SELECT 2699, 5 UNION ALL SELECT 2699, 6 UNION ALL SELECT 2699, 7 UNION ALL SELECT 2699, 8 UNION ALL SELECT 2699, 9 UNION ALL SELECT 2699, 10 UNION ALL SELECT 2699, 11) AS vals ON ((c.oid = vals.oid) AND (a.attnum = vals.attnum))`

`SELECT c.oid, a.attnum, a.attname, c.relname, n.nspname, a.attnotnull OR ((t.typtype = 'd') AND t.typnotnull), pg_get_expr(d.adbin, d.adrelid) LIKE '%nextval(%' FROM pg_catalog.pg_class AS c JOIN pg_catalog.pg_namespace AS n ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute AS a ON (c.oid = a.attrelid) JOIN pg_catalog.pg_type AS t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef AS d ON ((d.adrelid = a.attrelid) AND (d.adnum = a.attnum)) JOIN (SELECT 2690 AS oid, 1 AS attnum) AS vals ON ((c.oid = vals.oid) AND (a.attnum = vals.attnum))`

This is the output on postgres:

```
PostgreSQL 8.3.23

postgres=# SELECT c.oid, a.attnum, a.attname, c.relname, n.nspname, a.attnotnull OR ((t.typtype = 'd') AND t.typnotnull), pg_get_expr(d.adbin, d.adrelid) LIKE '%nextval(%' FROM pg_catalog.pg_class AS c JOIN pg_catalog.pg_namespace AS n ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute AS a ON (c.oid = a.attrelid) JOIN pg_catalog.pg_type AS t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef AS d ON ((d.adrelid = a.attrelid) AND (d.adnum = a.attnum)) JOIN (SELECT 2690 AS oid, 1 AS attnum) AS vals ON ((c.oid = vals.oid) AND (a.attnum = vals.attnum));
oid | attnum | attname | relname | nspname | ?column? | ?column?
------+--------+---------+-------------------+------------+----------+----------
2690 | 1 | oid | pg_proc_oid_index | pg_catalog | f |
(1 row)

postgres=# SELECT c.oid, a.attnum, a.attname, c.relname, n.nspname, a.attnotnull OR ((t.typtype = 'd') AND t.typnotnull), pg_get_expr(d.adbin, d.adrelid) LIKE '%nextval(%' FROM pg_catalog.pg_class AS c JOIN pg_catalog.pg_namespace AS n ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute AS a ON (c.oid = a.attrelid) JOIN pg_catalog.pg_type AS t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef AS d ON ((d.adrelid = a.attrelid) AND (d.adnum = a.attnum)) JOIN (SELECT 2699 AS oid, 1 AS attnum UNION ALL SELECT 2699, 2 UNION ALL SELECT 2699, 3 UNION ALL SELECT 2699, 4 UNION ALL SELECT 2699, 5 UNION ALL SELECT 2699, 6 UNION ALL SELECT 2699, 7 UNION ALL SELECT 2699, 8 UNION ALL SELECT 2699, 9 UNION ALL SELECT 2699, 10 UNION ALL SELECT 2699, 11) AS vals ON ((c.oid = vals.oid) AND (a.attnum = vals.attnum));
oid | attnum | attname | relname | nspname | ?column? | ?column?
------+--------+--------------+-------------------------------+------------+----------+----------
2699 | 1 | tgconstrname | pg_trigger_tgconstrname_index | pg_catalog | f |
```

These metadata queries come from  https://github.com/pgjdbc/pgjdbc/blob/01dbbac5b6c43e1079a463119e13c1bee74cbfdb/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSetMetaData.java#L187

We believe that we're running into this problem when trying to inspect the `pg_trigger_tgconstraint_index` table. (OID 2699) which currently isn't supported in CockroachDB.

cc @vy-ton and @rafiss 

gz#7518

### PR DESCRIPTION
**Pull request by rafiss**
fixes #60533 

Future work here is to treat OIDs as unsigned 32-bit integers, and to
match the PG semantics of comparing signed integers to unsigned OIDs.

Release note (bug fix): The previous implementation of integer-oid
comparison is incorrect when the integer is negative. This
incorrectness is now fixed by explicitly disallowing negative int
comparison with oids.

Release justification: fix for high-priority bug, which is also
low-risk.

### CODE REVIEW

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -2241,15 +2287,38 @@ var CmpOps = cmpOpFixups(map[ComparisonOperator]cmpOpOverload{
...
+			LeftType:  types.Int,
+			RightType: types.Oid,
+			Fn: func(ctx *EvalContext, left Datum, right Datum) (Datum, error) {
+				i := MustBeDInt(left)
+				if i < 0 {
+					return nil, errCannotCompareNegativeIntToOid
+				}
+				return cmpOpScalarLEFn(ctx, left, right)
+			},
+			Volatility: VolatilityLeakProof,
```
Comment (otan) [inline pkg/sql/sem/tree/eval.go:2307]: nit: this is now Immutable

File: pkg/sql/sem/tree/eval.go [hunk 2]
```diff
@@ -2241,15 +2287,38 @@ var CmpOps = cmpOpFixups(map[ComparisonOperator]cmpOpOverload{
...
+			LeftType:  types.Oid,
+			RightType: types.Int,
+			Fn: func(ctx *EvalContext, left Datum, right Datum) (Datum, error) {
+				i := MustBeDInt(right)
+				if i < 0 {
+					return nil, errCannotCompareNegativeIntToOid
+				}
+				return cmpOpScalarLEFn(ctx, left, right)
+			},
+			Volatility: VolatilityLeakProof,
```
Comment (otan) [inline pkg/sql/sem/tree/eval.go:2319]: nit: this is now Immutable

**Review by otan — COMMENTED** [review]
postgres seems to allow this comparison; any reason we are disallowing it?

also seems dangerous if people are relying on this somehow

**Comment by rafiss** [comment]
we implemented the comparison incorrectly, and it's a bit tricky and confusing to get right, so rather than return an incorrect result, we will return an error now. (we need to backport something to 20.2)

e.g. postgres:
```
rafiss@127:postgres> select (-1)::int = 4294967295::oid
+------------+
| ?column?   |
|------------|
| True       |
+------------+

rafiss@127:postgres> select (-1)::int4::oid
+------------+
| oid        |
|------------|
| 4294967295 |
+------------+
```

see the linked issue for details. and this [slack thread](https://cockroachlabs.slack.com/archives/C0168LW5THS/p1614009986002400) for even more details and discussion about other ideas

basically the problem is that we use a `uint32` for our vtable OIDs, but java stores OIDs in a signed int.

also, we never should have let OIDs in  CRDB be signed, which is kinda related.

**Review by otan — APPROVED** [review]
i see, ok. as long as immutability changes.
you may need to change the `VolatilityMatchesPostgres` related tests to include this in the blocklist

**Comment by rafiss** [comment]
let's do https://github.com/cockroachdb/cockroach/pull/61148 instead
